cpp/src/arrow/util/thread_pool.h

```diff
@@ -218,7 +220,8 @@ class ARROW_EXPORT ThreadPool : public Executor {
 
   ThreadPool();
 
-  Status SpawnReal(TaskHints hints, std::function<void()> task) override;
+  Status SpawnReal(TaskHints hints, std::function<void()> task,
+                   StopToken* = NULLPTR) override;
```

Comment:
Thanks for the heads up. This looks good, we don't yet use the Executor interface but will move to it once we start looking at upstreaming/open-sourcing our project.

Reply:
Why expose the stop token to the executor at all?  If I'm writing a custom executor what do I do with the stop token?  I see that the thread pool uses this to skip the task if it has been cancelled prior to scheduling but conceivably this could be achieved by wrapping the task as well in the same way that the TaskGroup::AppendReal does.  Do you envision certain classes of custom executor will want to do something different with the stop token?

Reply:
Maybe I answered my own question.  The executor might get some benefit from freeing the resources sooner rather than later (e.g. to keep queues down and more efficient) and the executor can choose to completely ignore the stop token without any consequence so it isn't like it is putting much burden on the implementor of the executor interface.

Reply:
Yes, the idea is that it gives more freedom to the executor like this.

Reply:
If the producer is creating the stop token wouldn't the correct interface be

Result<StopToken*> SpawnReal(TaskHints hints, std::function<void()> task);

The executor is responsible for creating and returning a stop token.  For example, in the existing model, how does the executor inform the caller that it is, in fact, cancellable.

Reply:
Well, the stop token is created in `Executor::Submit` since it's embedded in the Future.